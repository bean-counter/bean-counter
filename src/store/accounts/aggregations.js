/* eslint-disable no-console */
import types from './types'

export const calculateCurrentBalance = (account, transactions = []) => (
  transactions.reduce((balance, transaction) => {
    if (transaction.createdAt > account.openingBalanceDate) {
      return balance + transaction.amount
    }
    return balance
  }, account.openingBalance)
)

export const groupByInstitution = () => {
  return (dispatch, getState) => {
    const { byId, byInstitution } = getState().accounts
    // Group sorted accounts by institution
    // results in something like:
    // {
    //   TD: {accountIds: [id1, id2], balance: 100},
    //   BMO: {accountIds: [id3], balance: 200}
    // }
    const groupedAccountIds = Object.values(byId)
      // Sort accounts by name
      .sort((a, b) => (a.name > b.name))
      // Group them by institution
      .reduce((result, account) => {
        const {
          id,
          institution,
          groupId,
          currentBalance
        } = account
        const institutionData = result[institution] || { groups: {} }
        const institutionBalance = institutionData.balance || 0
        const accountGroups = institutionData.groups
        const accountGroupData = accountGroups[groupId] || {
          id: groupId,
          type: 'default',
          balance: 0,
          accountIds: []
        }
        const existingInstitutionData = byInstitution[institution] || { groups: {} }
        const existingAccountGroupData = existingInstitutionData.groups[groupId] || {
          id: groupId,
          type: 'default',
          balance: 0,
          accountIds: []
        }

        return {
          ...result,
          [institution]: {
            ...existingInstitutionData,
            balance: institutionBalance + currentBalance,
            groups: {
              ...existingInstitutionData.groups,
              [groupId]: {
                ...existingAccountGroupData,
                balance: accountGroupData.balance + currentBalance,
                accountIds: [...accountGroupData.accountIds, id]
              }
            }
          }
        }
      }, {})

    // But now the institutions are not sorted and we lost existing institution data
    const payload = Object.keys(groupedAccountIds)
      // Sort institutions by name
      .sort((a, b) => (a > b))
      // Merge new sorted institutions with existing institution data
      .reduce((result, institution) => ({
        ...result,
        [institution]: {
          ...byInstitution[institution],
          ...groupedAccountIds[institution]
        }
      }), {})

    dispatch({ type: types.UPDATE_GROUPS_BY_INSTITUTION, payload })
  }
}
