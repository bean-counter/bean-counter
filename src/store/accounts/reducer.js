/* eslint-disable no-case-declarations */
import types from './types'

export const initialState = {
  byId: {}, // {id1: {id: id1, name: 'account1', institution: 'TD'}, id2: {...}}
  byInstitution: {} // {TD: {accountIds: [id1, id2], balance: 100, apiKey: XYZ}, BMO: {...}}
}


let accounts
let institutionData
export default (state = initialState, { type, payload }) => {
  switch (type) {
    case types.LOAD_ACCOUNTS:
      return payload || initialState
    case types.CREATE_ACCOUNT:
      accounts = { ...state.byId, [payload.id]: payload }
      return {
        ...state,
        byId: accounts
      }
    case types.UPDATE_ACCOUNT:
      accounts = { ...state.byId, [payload.id]: payload }
      return {
        ...state,
        byId: accounts
      }
    case types.DELETE_ACCOUNT:
      const { [payload]: _, ...rest } = state.byId
      return {
        ...state,
        byId: rest
      }
    case types.UPDATE_GROUPS_BY_INSTITUTION:
      return {
        ...state,
        byInstitution: payload
      }
    case types.CREATE_ACCOUNT_GROUP:
    case types.UPDATE_ACCOUNT_GROUP:
      institutionData = state.byInstitution[payload.institution] || { groups: {} }
      return {
        ...state,
        byInstitution: {
          ...state.byInstitution,
          [payload.institution]: {
            ...institutionData,
            groups: {
              ...institutionData.groups,
              [payload.accountGroup.id]: payload.accountGroup
            }
          }
        }
      }
    default:
      return state
  }
}
