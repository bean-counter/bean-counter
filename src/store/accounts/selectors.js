/* eslint no-console: 0 */
import { createSelector } from 'reselect'

const getAccounts = state => state.accounts
const getInstitutions = () => ([
  'RBC', 'BMO', 'TD', 'Tangerine'
])

const findAccountById = (state, { match }) => (
  state.accounts.find(account => account.id === match.params.accountId)
)

export const makeFindAccountById = () => createSelector(
  findAccountById,
  account => ({ account })
)

export const sortedAccountsGroupedByInstitution = createSelector(
  getAccounts,
  (accounts) => {
    const groupedAccounts = {}
    accounts.sort((a, b) => (a.name > b.name)).forEach((account) => {
      if (!Object.keys(groupedAccounts).includes(account.institution)) {
        groupedAccounts[account.institution] = []
      }
      return groupedAccounts[account.institution].push(account)
    })
    return groupedAccounts
  }
)


export const sortedInstitutions = createSelector(
  getInstitutions,
  institutions => institutions.sort((a, b) => (a > b))
)

export const sortedInstitutionsForAutoselect = createSelector(
  sortedInstitutions,
  institutions => institutions.map(institution => ({ label: institution, value: institution }))
)
