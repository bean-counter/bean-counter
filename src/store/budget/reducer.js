/* eslint-disable no-case-declarations */
import uuid from 'uuid/v4'
import types from './types'
import budgetCategories from '../../data/budgetCategories'

// https://projects.susielu.com
// http://repec.sowi.unibe.ch/stata/palettes/index.html
const defaultColours = [
  '#1f78b4', '#b2df8a', '#e31a1c',
  '#ff7f00', '#cab2d6', '#a6cee3',
  '#33a02c', '#6a3d9a', '#fb9a99',
  '#fdbf6f', '#ffff99', '#b15928'
]


const categoryTree = (categoriesById) => {
  const parents = Object.values(categoriesById).reduce((result, parentCategory) => {
    if (!('parentId' in parentCategory)) {
      return {
        ...result,
        [parentCategory.id]: {
          id: parentCategory.id,
          label: parentCategory.name,
          options: []
        }
      }
    }
    return result
  }, {})
  Object.values(categoriesById).forEach((category) => {
    if (category.parentId in parents) {
      parents[category.parentId].options.push({
        id: category.id,
        label: category.name,
        value: category.name,
        colour: category.colour
      })
    }
  })
  return Object.values(parents)
}

const generateInitialState = () => {
  let count = 0
  const categoriesById = Object.keys(budgetCategories).reduce((result, category) => {
    const parentId = uuid()
    const parent = {
      id: parentId,
      name: category
    }
    const children = budgetCategories[category].reduce((res, subCategory) => {
      const child = {
        id: uuid(),
        parentId,
        name: subCategory,
        colour: defaultColours[count % defaultColours.length]
      }
      count += 1
      return { ...res, [child.id]: child }
    }, {})

    return {
      ...result,
      [parent.id]: parent,
      ...children
    }
  }, {})

  return {
    categoriesById,
    categoryTree: categoryTree(categoriesById),
    rules: {} // rules have the format => match: { category: 'cat 1', type: 'exact_match', count: 0 }
  }
}
export const initialState = generateInitialState()
let categoriesById

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case types.LOAD_BUDGET:
      return payload || initialState
    case types.CREATE_CATEGORY:
      const topCategory = state.categoryTree.find((cat => cat.id === payload.category.parentId))
      const lastColour = topCategory.options[topCategory.options.length - 1].colour
      const lastColourIndex = defaultColours.indexOf(lastColour)
      categoriesById = {
        ...state.categoriesById,
        [payload.category.id]: {
          ...payload.category,
          colour: defaultColours[(lastColourIndex + 1) % defaultColours.length]
        }
      }
      return {
        ...state,
        categoriesById,
        categoryTree: categoryTree(categoriesById)
      }
    case types.UPDATE_CATEGORY:
      categoriesById = { ...state.categoriesById, [payload.id]: payload }
      return {
        ...state,
        categoriesById,
        categoryTree: categoryTree(categoriesById)
      }
    case types.DELETE_CATEGORY:
      const { [payload]: category, ...rest } = state.categoriesById
      return {
        ...state,
        categoriesById: rest,
        categoryTree: categoryTree(rest)
      }
    case types.CREATE_EXACT_RULE:
      return {
        ...state,
        rules: {
          ...state.rules,
          [payload.match]: {
            categoryId: payload.categoryId,
            type: 'exact_match',
            count: 0
          }
        }
      }
    case types.DELETE_EXACT_RULE:
      const r = Object.keys(state.rules).reduce((result, match) => {
        if (match !== payload) {
          return {
            ...result,
            rules: { ...result.rules, [match]: state.rules[match] }
          }
        }
        return result
      }, { ...state, rules: {} })
      return r
    case types.COUNT_RULE_USAGE:
      // reset counters
      const newState = Object.keys(state.rules).reduce((result, match) => (
        {
          ...result,
          rules: {
            ...result.rules,
            [match]: { ...result.rules[match], count: 0 }
          }
        }
      ), state)

      // count the transactions
      for (const transaction of payload) {
        if (transaction.description in newState.rules) {
          newState.rules[transaction.description].count += 1
        }
      }
      // remove unused rules
      return Object.keys(newState.rules).reduce((result, match) => {
        if (newState.rules[match].count > 0) {
          return {
            ...result,
            rules: { ...result.rules, [match]: newState.rules[match] }
          }
        }
        return result
      }, { ...newState, rules: {} })
    default:
      return state
  }
}
