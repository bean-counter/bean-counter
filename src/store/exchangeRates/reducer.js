import _ from 'lodash'
import types from './types'

export const initialState = {}

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_EXCHANGE_RATES:
      return action.payload || initialState
    case types.CREATE_EXCHANGE_RATE:
      return {
        ...state,
        [action.payload.currency]: {
          exchangeRate: action.payload.exchangeRate,
          updatedOn: action.payload.updatedOn
        }
      }
    case types.UPDATE_EXCHANGE_RATE:
      return {
        ...state,
        [action.payload.currency]: {
          exchangeRate: action.payload.exchangeRate,
          updatedOn: action.payload.updatedOn
        }
      }
    case types.DELETE_CURRENCIES:
      return _.omit(state, action.payload)
    default:
      return state
  }
}
