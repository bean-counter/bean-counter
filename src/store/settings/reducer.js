import _ from 'lodash'
import types from './types'

export const initialState = {
  selectedAccountId: null,
  portfolioFilters: {
    institution: {},
    account: {}
  }
}

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_SETTINGS:
      return action.payload || initialState
    case types.CREATE_PORTFOLIO_FILTERS:
      return {
        ...state,
        portfolioFilters: {
          ...state.portfolioFilters,
          [action.payload.filterName]: {
            ...state.portfolioFilters[action.payload.filterName],
            ...action.payload.options
          }
        }
      }
    case types.DELETE_PORTFOLIO_FILTERS:
      return {
        ...state,
        portfolioFilters: {
          ...state.portfolioFilters,
          [action.payload.filterName]: _.omit(state.portfolioFilters[action.payload.filterName], action.payload.options)
        }
      }
    case types.UPDATE_PORTFOLIO_FILTER_VALUE:
      return {
        ...state,
        portfolioFilters: {
          ...state.portfolioFilters,
          [action.payload.filterName]: {
            ...state.portfolioFilters[action.payload.filterName],
            [action.payload.option]: action.payload.value
          }
        }
      }
    case types.SELECT_ACCOUNT:
      return {
        ...state,
        selectedAccountId: action.payload
      }
    default:
      return state
  }
}
