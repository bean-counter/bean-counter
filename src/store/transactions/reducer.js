import types from './types'

export const initialState = {
  transactions: []
}

export default (state = initialState, action) => {
  switch (action.type) {
    case types.ADD_TRANSACTION:
      return { ...state, transactions: state.transactions.concat(action.payload) }
    default:
      return state
  }
}
