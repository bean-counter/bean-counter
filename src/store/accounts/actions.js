/* eslint-disable no-console */
import uuid from 'uuid/v4'
import types from './types'
import { saveState } from '../user/actions'
import { showOverlay, hideOverlay } from '../settings/actions'
import { calculateCurrentBalance, groupByInstitution } from './aggregations'
import {
  addTransactions,
  deleteTransactions,
  getAccountTransactions
} from '../transactions/actions'

export const loadAccounts = (accounts) => {
  return { type: types.LOAD_ACCOUNTS, payload: accounts }
}

export const afterAccountsChanged = async (dispatch) => {
  await dispatch(groupByInstitution())
  saveState()
}

export const createAccount = (account) => {
  return async (dispatch) => {
    const newAccount = {
      ...account,
      groupId: account.groupId || '0',
      currentBalance: calculateCurrentBalance(account),
      id: uuid()
    }
    dispatch({ type: types.CREATE_ACCOUNT, payload: newAccount })
    await afterAccountsChanged(dispatch)
    return newAccount.id
  }
}

export const updateAccount = (account) => {
  return (dispatch, getState) => {
    const transactions = getAccountTransactions(account.id, getState().transactions.list)
    const currentBalance = calculateCurrentBalance(account, transactions)
    dispatch({ type: types.UPDATE_ACCOUNT, payload: { ...account, currentBalance } })
    afterAccountsChanged(dispatch)
  }
}

export const updateAccountBalance = (account) => {
  return (dispatch, getState) => {
    const transactions = getAccountTransactions(account.id, getState().transactions.list)
    const currentBalance = calculateCurrentBalance(account, transactions)
    dispatch({ type: types.UPDATE_ACCOUNT, payload: { ...account, currentBalance } })
  }
}

export const deleteAccount = (account, options = { skipAfterChange: false }) => {
  const { skipAfterChange } = options
  return (dispatch, getState) => {
    const transactionIds = getState().transactions.list
      .filter(transaction => (
        transaction.accountId === account.id
      ))
      .map(transaction => transaction.id)

    dispatch(deleteTransactions(account, transactionIds, { skipAfterChange: true }))
    dispatch({ type: types.DELETE_ACCOUNT, payload: account.id })
    if (!skipAfterChange) {
      afterAccountsChanged(dispatch)
    }
  }
}

export const createAccountWithTransactions = async (dispatch, account, transactions) => {
  const newAccount = {
    ...account,
    groupId: account.groupId || '0',
    currentBalance: calculateCurrentBalance(account, transactions),
    id: uuid()
  }
  const newTransactions = transactions.map(transaction => ({
    ...transaction,
    id: uuid(),
    accountId: newAccount.id
  }))

  await dispatch({ type: types.CREATE_ACCOUNT, payload: newAccount })
  await dispatch(addTransactions(newAccount, newTransactions, { skipAfterChange: true }))
  return newAccount.id
}

export const createAccountGroup = (institution, accountGroupData, importedAccounts) => {
  return async (dispatch) => {
    dispatch(showOverlay(`Importing data from ${institution} ...`))
    const accountGroupId = uuid()

    const accountIds = await Promise.all(await importedAccounts.map(async (importedAccount) => {
      const { transactions, ...newAccount } = importedAccount
      newAccount.institution = institution
      newAccount.groupId = accountGroupId
      return createAccountWithTransactions(dispatch, newAccount, transactions)
    }))

    // Create the account group
    const accountGroup = {
      ...accountGroupData,
      accountIds,
      id: accountGroupId,
      type: 'api'
    }

    await dispatch({ type: types.CREATE_ACCOUNT_GROUP, payload: { institution, accountGroup } })
    await saveState()
    dispatch(hideOverlay())
  }
}

export const updateAccountGroup = (institution, accountGroup, importedAccounts) => {
  return async (dispatch, getState) => {
    dispatch(showOverlay(`Importing data from ${institution} ...`))

    const existingAccounts = []
    await Promise.all(importedAccounts.map(async (importedAccount) => {
      // Find existing account
      const account = Object.values(getState().accounts.byId)
        .find(acc => acc.sourceId === importedAccount.sourceId)

      if (account) {
        existingAccounts.push(account)
        // TODO: add new transactions
      } else {
        const { transactions, ...newAccount } = importedAccount
        newAccount.institution = institution
        newAccount.groupId = accountGroup.id
        return createAccountWithTransactions(dispatch, newAccount, transactions)
      }
      return null
    }))
    return dispatch(hideOverlay())
  }
}

export const deleteAccountGroup = (accountGroup) => {
  return async (dispatch, getState) => {
    Object.values(accountGroup.accountIds).map((accountId) => {
      return dispatch(deleteAccount(getState.accounts[accountId], { skipAfterChange: true }))
    })
    afterAccountsChanged(dispatch)
  }
}
