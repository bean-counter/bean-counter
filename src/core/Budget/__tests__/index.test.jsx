import React from 'react'
import { mount } from 'enzyme'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import BudgetIndex from '..'
import { initialState as accountsInitialState } from '../../../store/accounts/reducer'
import { initialState as transactionsInitialState } from '../../../store/transactions/reducer'
import { initialState as budgetInitialState } from '../../../store/budget/reducer'
import { initialState as settingsInitialState } from '../../../store/settings/reducer'

describe('BudgetIndex', () => {
  it('matches snapshot with no accounts', () => {
    const mockStore = configureMockStore()
    const store = mockStore({
      accounts: accountsInitialState,
      transactions: transactionsInitialState,
      budget: budgetInitialState,
      settings: settingsInitialState
    })
    const wrapper = mount((
      <Provider store={store}>
        <BudgetIndex />
      </Provider>
    ))
    expect(wrapper.debug()).toMatchSnapshot()
  })
})
