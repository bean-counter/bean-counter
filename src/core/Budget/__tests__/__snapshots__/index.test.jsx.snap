// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BudgetIndex matches snapshot with no accounts 1`] = `
"<Provider store={{...}}>
  <BudgetIndex>
    <WithStyles(ForwardRef(Container)) className=\\"makeStyles-root-1\\">
      <ForwardRef(Container) classes={{...}} className=\\"makeStyles-root-1\\">
        <div className=\\"MuiContainer-root makeStyles-root-1 MuiContainer-maxWidthLg\\">
          <WithStyles(ForwardRef(Grid)) container={true} spacing={3}>
            <ForwardRef(Grid) classes={{...}} container={true} spacing={3}>
              <div className=\\"MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3\\">
                <WithStyles(ForwardRef(Grid)) item={true} xs={12} className=\\"makeStyles-pageHeader-3\\">
                  <ForwardRef(Grid) classes={{...}} item={true} xs={12} className=\\"makeStyles-pageHeader-3\\">
                    <div className=\\"MuiGrid-root makeStyles-pageHeader-3 MuiGrid-item MuiGrid-grid-xs-12\\">
                      <WithStyles(ForwardRef(Typography)) variant=\\"h5\\">
                        <ForwardRef(Typography) classes={{...}} variant=\\"h5\\">
                          <h5 className=\\"MuiTypography-root MuiTypography-h5\\">
                            Budget history
                          </h5>
                        </ForwardRef(Typography)>
                      </WithStyles(ForwardRef(Typography))>
                      <div className=\\"makeStyles-filters-4\\">
                        <PopupDateRangePicker ranges={{...}} onChange={[Function: handleSelectDate]} minDate={{...}} maxDate={{...}}>
                          <WithStyles(ForwardRef(Button)) aria-describedby={[undefined]} variant=\\"contained\\" onClick={[Function: handleOpenCalendar]} endIcon={{...}}>
                            <ForwardRef(Button) classes={{...}} aria-describedby={[undefined]} variant=\\"contained\\" onClick={[Function: handleOpenCalendar]} endIcon={{...}}>
                              <WithStyles(ForwardRef(ButtonBase)) className=\\"MuiButton-root MuiButton-contained\\" component=\\"button\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"Mui-focusVisible\\" type=\\"button\\" aria-describedby={[undefined]} onClick={[Function: handleOpenCalendar]}>
                                <ForwardRef(ButtonBase) classes={{...}} className=\\"MuiButton-root MuiButton-contained\\" component=\\"button\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"Mui-focusVisible\\" type=\\"button\\" aria-describedby={[undefined]} onClick={[Function: handleOpenCalendar]}>
                                  <button className=\\"MuiButtonBase-root MuiButton-root MuiButton-contained\\" onBlur={[Function]} onClick={[Function: handleOpenCalendar]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={0} type=\\"button\\" disabled={false} aria-describedby={[undefined]}>
                                    <span className=\\"MuiButton-label\\">
                                      Aug 01, 2019
                                       -Â 
                                      Nov 13, 2019
                                      <span className=\\"MuiButton-endIcon MuiButton-iconSizeMedium\\">
                                        <ForwardRef>
                                          <WithStyles(ForwardRef(SvgIcon))>
                                            <ForwardRef(SvgIcon) classes={{...}}>
                                              <svg className=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" color={[undefined]} aria-hidden=\\"true\\" role=\\"presentation\\">
                                                <path d=\\"M20 3h-1V1h-2v2H7V1H5v2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 18H4V8h16v13z\\" />
                                              </svg>
                                            </ForwardRef(SvgIcon)>
                                          </WithStyles(ForwardRef(SvgIcon))>
                                        </ForwardRef>
                                      </span>
                                    </span>
                                    <NoSsr>
                                      <WithStyles(undefined) center={false}>
                                        <ForwardRef(TouchRipple) classes={{...}} center={false}>
                                          <span className=\\"MuiTouchRipple-root\\">
                                            <TransitionGroup component={{...}} exit={true} childFactory={[Function: childFactory]} />
                                          </span>
                                        </ForwardRef(TouchRipple)>
                                      </WithStyles(undefined)>
                                    </NoSsr>
                                  </button>
                                </ForwardRef(ButtonBase)>
                              </WithStyles(ForwardRef(ButtonBase))>
                            </ForwardRef(Button)>
                          </WithStyles(ForwardRef(Button))>
                          <WithStyles(ForwardRef(Popover)) id={[undefined]} open={false} anchorEl={{...}} onClose={[Function: handleCloseCalendar]} anchorOrigin={{...}} transformOrigin={{...}}>
                            <ForwardRef(Popover) classes={{...}} id={[undefined]} open={false} anchorEl={{...}} onClose={[Function: handleCloseCalendar]} anchorOrigin={{...}} transformOrigin={{...}}>
                              <ForwardRef(Modal) container={[undefined]} open={false} BackdropProps={{...}} className=\\"MuiPopover-root\\" id={[undefined]} onClose={[Function: handleCloseCalendar]} />
                            </ForwardRef(Popover)>
                          </WithStyles(ForwardRef(Popover))>
                        </PopupDateRangePicker>
                      </div>
                    </div>
                  </ForwardRef(Grid)>
                </WithStyles(ForwardRef(Grid))>
                <WithStyles(ForwardRef(Grid)) item={true} xs={12}>
                  <ForwardRef(Grid) classes={{...}} item={true} xs={12}>
                    <div className=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12\\">
                      <WithStyles(ForwardRef(Paper)) className=\\"makeStyles-budgetChart-2\\">
                        <ForwardRef(Paper) classes={{...}} className=\\"makeStyles-budgetChart-2\\">
                          <div className=\\"MuiPaper-root MuiPaper-elevation1 makeStyles-budgetChart-2 MuiPaper-rounded\\">
                            <BudgetLineChart data={{...}} categories={{...}} onLegendClick={[Function: handleLegendClick]}>
                              <ResponsiveContainer width=\\"100%\\" height=\\"100%\\" debounce={0}>
                                <div id={[undefined]} className=\\"recharts-responsive-container\\" style={{...}}>
                                  <LineChart data={{...}} margin={{...}} layout=\\"horizontal\\" stackOffset=\\"none\\" barCategoryGap=\\"10%\\" barGap={4} reverseStackOrder={false} width={0} height={0} />
                                  <ResizeDetector handleWidth={true} handleHeight={true} onResize={[Function]} skipOnMount={false} refreshRate={1000} refreshMode={[undefined]} resizableElementId=\\"\\">
                                    <div style={{...}} />
                                  </ResizeDetector>
                                </div>
                              </ResponsiveContainer>
                            </BudgetLineChart>
                          </div>
                        </ForwardRef(Paper)>
                      </WithStyles(ForwardRef(Paper))>
                    </div>
                  </ForwardRef(Grid)>
                </WithStyles(ForwardRef(Grid))>
              </div>
            </ForwardRef(Grid)>
          </WithStyles(ForwardRef(Grid))>
        </div>
      </ForwardRef(Container)>
    </WithStyles(ForwardRef(Container))>
  </BudgetIndex>
</Provider>"
`;
